### PR TITLE
Track IsEffectivelyVisible state

### PR DESCRIPTION
**Pull request by jmacato**
Previously, `IsEffectivelyVisible` was being calculated each time it was called. This was problematic for a couple of reasons:

1. It traversed the tree each time it was called
2. There is no way to track changes to `IsEffectivelyVisible`

This PR changes `IsEffectivelyVisible` to be tracked like `IsEffectivelyEnabled` which fixes 1). It doesn't add a way to track `IsEffectivelyVisible` changes yet though so doesn't address 2).

Regarding 2), making a `IsEffectivelyVisible` a `DirectProperty` may be problematic for performance reasons as the change notifications need to be fired even if there are no listeners. A CLR event may be acceptable, but the main reason we need to track changes to this is internally for the animation system (PR to follow which hooks into this) so there are currently no plans to expose an API for detecting changes to this.

### CODE REVIEW

**Review by jp2masa — COMMENTED** [review]
Some method names seem to be wrong: `UpdateIsEffectivelyEnabled`, `SyncIsEffectivelyEnabledFromParent`, it should be `Visible`, not `Enabled`.

Also, I may be reading it wrong, but not sure if the logic always works, e.g. does adding this to the test work?

```cs
child2.IsVisible = false;
root.IsVisible = true;

Assert.False(child2.IsEffectivelyEnabled);
```

Finally, I think there's room for optimization in `UpdateIsEffectivelyEnabled`, so that it only needs to do the recursive call when `child.IsVisible`.

**Comment by jmacato** [comment]
@jp2masa thanks for the advise! 

>child2.IsVisible = false;
>root.IsVisible = true;
>Assert.False(child2.IsEffectivelyEnabled);

Good catch on this one! I'm adding this to the test now!

> Finally, I think there's room for optimization in UpdateIsEffectivelyEnabled, so that it only needs to do the recursive call when child.IsVisible.

Makes sense

**Comment by workgroupengineering** [comment]
We could simplify the code this way by avoiding loops.

```c#
        /// <summary>
        /// Gets a value indicating whether this control and all its parents are visible.
        /// </summary>
        public bool IsEffectivelyVisible => 
            IsVisible && (VisualParent?.IsEffectivelyVisible ?? true);
```

**Comment by jmacato** [comment]
@workgroupengineering we intend to provide internal event updates when IEV changes, so we can't do what you suggested there unfortunately.

**Review by grokys — CHANGES_REQUESTED** [review]
A few nits, but also this doesn't seem to handle correctly setting the `IsEffectivelyVisible` state when adding a grandchild. Here is a failing test:

```
        [Fact]
        public void Added_Grandchild_Has_Correct_IsEffectivelyVisible()
        {
            var child = new Decorator();
            var grandchild = new Decorator();
            var root = new TestRoot 
            { 
                IsVisible = false,
                Child = child
            };

            child.Child = grandchild;
            Assert.False(grandchild.IsEffectivelyVisible);
        }
```

File: src/Avalonia.Base/Visual.cs [hunk 1]
```diff
@@ -495,10 +515,13 @@ protected virtual void OnAttachedToVisualTreeCore(VisualTreeAttachmentEventArgs
...
+            
             _visualRoot.Renderer.RecalculateChildren(_visualParent!);
-
-            if (ZIndex != 0 && VisualParent is Visual parent)
-                parent.HasNonUniformZIndexChildren = true;
+            
+            if (ZIndex != 0 && _visualParent is { })
+                _visualParent.HasNonUniformZIndexChildren = true;
+            
+            SyncIsEffectivelyVisible(_visualParent);
```
Comment (grokys) [inline src/Avalonia.Base/Visual.cs:524]: Question: should `IsEffectivelyVisible` be set when the `AttachedToVisualTree`/`DetachedFromVisualTree` event is called? My initial thought is "yes", though you may know more than me.

**Comment by grokys** [comment]
The `VisualExtensions_GetVisualsAt.Should_Find_Control` test seems to be flaky: I've seen it occur on macOS and now the Linux builds, but it passes locally. Not sure if it's caused by this PR or not.

**Review by grokys — APPROVED** [review]

**Pull request merged**